spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/TimefoldBeanFactory.java

```diff
@@ -0,0 +1,204 @@
+        failInjectionWithMultipleSolvers(ScoreManager.class.getName());
+        SolverFactory solverFactory = context.getBean(SolverFactory.class);
+        if (solverFactory == null) {
+            return null;
+        }
+        return ScoreManager.create(solverFactory);
+    }
+
+    @Bean
+    @Lazy
+    @ConditionalOnMissingBean
+    public <Solution_, Score_ extends Score<Score_>> SolutionManager<Solution_, Score_> solutionManager() {
+        failInjectionWithMultipleSolvers(SolutionManager.class.getName());
+        SolverFactory solverFactory = context.getBean(SolverFactory.class);
+        if (solverFactory == null) {
+            return null;
+        }
+        return SolutionManager.create(solverFactory);
+    }
+
+    // @Bean wrapped by static class to avoid classloading issues if dependencies are absent
+    @ConditionalOnClass({ ConstraintVerifier.class })
+    @ConditionalOnMissingBean({ ConstraintVerifier.class })
+    @AutoConfigureAfter(TimefoldAutoConfiguration.class)
+    class TimefoldConstraintVerifierConfiguration {
+
+        private final ApplicationContext context;
+
+        protected TimefoldConstraintVerifierConfiguration(ApplicationContext context) {
+            this.context = context;
+        }
+
+        @Bean
+        @Lazy
+        @SuppressWarnings("unchecked")
+        <ConstraintProvider_ extends ConstraintProvider, SolutionClass_>
+                ConstraintVerifier<ConstraintProvider_, SolutionClass_> constraintVerifier() {
+            // Using SolverConfig as an injected parameter here leads to an injection failure on an empty app,
+            // so we need to get the SolverConfig from context
+            failInjectionWithMultipleSolvers(ConstraintProvider.class.getName());
+            SolverConfig solverConfig;
+            try {
+                solverConfig = context.getBean(SolverConfig.class);
+            } catch (BeansException exception) {
+                solverConfig = null;
+            }
+
+            ScoreDirectorFactoryConfig scoreDirectorFactoryConfig =
+                    (solverConfig != null) ? solverConfig.getScoreDirectorFactoryConfig() : null;
+            if (scoreDirectorFactoryConfig == null || scoreDirectorFactoryConfig.getConstraintProviderClass() == null) {
+                // Return a mock ConstraintVerifier so not having ConstraintProvider doesn't crash tests
+                // (Cannot create custom condition that checks SolverConfig, since that
+                //  requires TimefoldAutoConfiguration to have a no-args constructor)
+                final String noConstraintProviderErrorMsg = (scoreDirectorFactoryConfig != null)
+                        ? "Cannot provision a ConstraintVerifier because there is no ConstraintProvider class."
+                        : "Cannot provision a ConstraintVerifier because there is no PlanningSolution or PlanningEntity classes.";
+                return new ConstraintVerifier<>() {
```

Comment:
Let's make this a proper class, inner if you prefer; reads better in stack traces and profiler results.